src/CompariosRelations.dfy

```diff
@@ -0,0 +1,44 @@
+include "Helpers.dfy"
+
+module Relations.ComparisonsRelation { 
+
+    export 
+    reveals Reflexive, AntiSymmetric, Connected, TotalOrdering
+    import opened Helpers.Comparison
+
+
+    predicate Reflexive<T(!new)>(R: (T, T) -> CompResult) 
+    {
+        forall x :: R(x, x)
```

Comment:
How can this typecheck if R returns a CompResult?

Reply:
This raises the interesting question that these relations are awkward to define (if possible at all) for relations that return `CompResult` (and therefore aren't really relations in the normal mathematical sense). I tried rewriting `Trichotomous` for `CompResult` to see what it'd look like (since I think it may be a sufficient precondition for both searching and sorting) and came up with this, which feels like a bit of a mess in relation to the original:

```
predicate Trichotomous<T(!new)>(compare: (T, T) -> CompResult) 
{
    (forall x, y :: compare(x, y).LessThan? || compare(x, y).Equals? || compare(y, x).LessThan?) &&  // at least one of the three using less
    (forall x, y :: compare(x, y).GreaterThan? || compare(x, y).Equals? || compare(y, x).GreaterThan?) &&  // at least one of the three using greater
    (forall x, y :: compare(x, y).LessThan? && compare(y, x).LessThan? ==> false) &&  // not both less
    (forall x, y :: compare(x, y).GreaterThan? && compare(y, x).GreaterThan? ==> false) &&  // not both greater
    (forall x, y :: compare(x, y).Equals? ==> x == y) && // Equals means equality
    (forall x, y :: compare(x, y).LessThan? ==> x != y) && // not less and equal
    (forall x, y :: compare(x, y).GreaterThan? ==> x != y)  // not greater and equal
}
```

Maybe using a less-than relation is actually a better approach in Dafny, given how well-behaved it is mathematically?

Reply:
Alternatively, we could adopt all of the code [from here](https://github.com/dafny-lang/compiler-bootstrap/blob/cpitclaudel_entities-impl/src/Utils/Library.dfy#L642-L748) directly.

Reply:
Yes, having considered it more, let's adopt that `compiler-bootstrap` code unless it leads to other issues.

Reply:
Yep, defining `.Le` and taking about it is the way to go, I think.
But note that in general for sorting you don't need to know that `Equals` implies `==` (and, in fact, you don't want it, either, since you may sort according to a projection of the value like `function Le(a: (int, int), b: (int, int)) { a[0] <= b[0] }`)

Reply:
Yep, both good points (and further arguments for using your code!).